packages/server/src/sdk/app/tables/internal/index.ts

```diff
@@ -0,0 +1,187 @@
+  const tableSaveFunctions = new TableSaveFunctions({
+    user: opts?.user,
+    oldTable,
+    importRows: opts?.rowsToImport,
+  })
+  table = await tableSaveFunctions.before(table)
+
+  // make sure that types don't change of a column, have to remove
+  // the column if you want to change the type
+  if (oldTable && oldTable.schema) {
+    for (const propKey of Object.keys(table.schema)) {
+      let oldColumn = oldTable.schema[propKey]
+      if (oldColumn && oldColumn.type === FieldTypes.INTERNAL) {
+        oldTable.schema[propKey].type = FieldTypes.AUTO
+      }
+    }
+  }
+
+  let renaming = opts?.renaming
+  if (renaming && renaming.old === renaming.updated) {
+    renaming = undefined
+  }
+
+  // rename row fields when table column is renamed
+  if (renaming && table.schema[renaming.updated].type === FieldTypes.LINK) {
+    throw new Error("Cannot rename a linked column.")
+  }
+
+  table = await tableSaveFunctions.mid(table)
+
+  // update schema of non-statistics views when new columns are added
+  for (let view in table.views) {
+    const tableView = table.views[view]
+    if (!tableView) continue
+
+    if (viewsSdk.isV2(tableView)) {
+      table.views[view] = viewsSdk.syncSchema(
+        oldTable!.views![view] as ViewV2,
+        table.schema,
+        renaming
+      )
+      continue
+    }
+
+    if (
+      (tableView.schema as ViewStatisticsSchema).group ||
+      tableView.schema.field
+    )
+      continue
+    tableView.schema = table.schema
+  }
+
+  // update linked rows
+  try {
+    const linkResp: any = await updateLinks({
+      eventType: oldTable ? EventType.TABLE_UPDATED : EventType.TABLE_SAVE,
+      table: table,
+      oldTable: oldTable,
+    })
+    if (linkResp != null && linkResp._rev) {
+      table._rev = linkResp._rev
+    }
+  } catch (err) {
+    throw new Error(err as string)
+  }
+
+  // don't perform any updates until relationships have been
+  // checked by the updateLinks function
+  const updatedRows = tableSaveFunctions.getUpdatedRows()
+  if (updatedRows && updatedRows.length !== 0) {
+    await db.bulkDocs(updatedRows)
+  }
+  let result = await db.put(table)
+  table._rev = result.rev
+  const savedTable = cloneDeep(table)
+
+  table = await tableSaveFunctions.after(table)
+  // the table may be updated as part of the table save after functionality - need to write it
+  if (!isEqual(savedTable, table)) {
+    result = await db.put(table)
+    table._rev = result.rev
+  }
+  // has to run after, make sure it has _id
+  await runStaticFormulaChecks(table, { oldTable, deletion: false })
+  return { table }
+}
+
+export async function destroy(table: Table) {
+  const db = context.getAppDB()
+  const tableId = table._id!
+
+  // Delete all rows for that table
+  const rowsData = await db.allDocs(
+    getRowParams(tableId, null, {
+      include_docs: true,
+    })
+  )
+  await db.bulkDocs(
+    rowsData.rows.map((row: any) => ({ ...row.doc, _deleted: true }))
```

Comment:
I've seen writes that set `_deleted: true` before, is this a soft delete or is this how you signal to Couch to do a hard delete?

Reply:
This is how you do a bulk delete, you can either call a specific `DELETE` endpoint for a single document, or you can write to the bulk documents API, with the documents tagged with a `_deleted` field.